test/services/task_list_merge_test.dart

```diff
@@ -0,0 +1,189 @@
+      (await devices[1].taskListService.getTaskListById('id2'))?.elements,
+      [task2],
+    );
+    expect(await devices[1].taskListService.getTasksFromList('id2'), [task2]);
+  });
+
+  test('crdt merge tasks unordered same list', () async {
+    final task1 = Task(title: 'task1');
+    final task2 = Task(title: 'task2');
+    await devices[0]
+        .taskListService
+        .upsertTaskList(TaskList(id: 'id', title: 'list'));
+    await devices[0].taskListService.upsertTask('id', task1);
+    await devices[1]
+        .taskListService
+        .upsertTaskList(TaskList(id: 'id', title: 'list'));
+    await devices[1].taskListService.upsertTask('id', task2);
+
+    await devices[0]
+        .taskListService
+        .mergeCrdtJson(await devices[1].taskListService.crdtToJson());
+
+    final taskLists = (await devices[0].taskListService.taskLists).toList();
+    expect(taskLists.length, 1);
+    expect(taskLists.first.id, 'id');
+    expect(taskLists.first.title, 'list');
+    expect(taskLists.first.elements.toSet(), {task1, task2});
+    expect(
+      (await devices[0].taskListService.getTasksFromList('id')).toSet(),
+      {task1, task2},
+    );
+    expect((await devices[0].taskListService.allTasks).toSet(), {task1, task2});
+  });
+
+  test('crdt merge clock drift', () async {
+    // This test is highly dependent on the underlying crdt implementation.
+    // The flutter crdt library throws a ClockDriftException when trying to merge with a Hlc timestamp that is more than 1 minute in the future.
+    // The new implementation should not care.
+
+    final taskList1 = TaskList(id: 'id1', title: 'taskList1');
+    final taskList2 = TaskList(id: 'id2', title: 'taskList2');
+    await devices[0].taskListService.upsertTaskList(taskList1);
+    await devices[1].taskListService.upsertTaskList(taskList2);
+
+    final Map<String, dynamic> device1Crdt =
+        jsonDecode(await devices[1].taskListService.crdtToJson());
+    final Map<String, dynamic> device1Tasks = device1Crdt['records'];
+    expect(device1Tasks.length, 1);
+    final Map<String, dynamic> device1Task1 = device1Tasks.values.first;
+    expect(device1Task1.containsKey('clock'), true);
+    expect(device1Task1['clock'].containsKey('timestamp'), true);
+    device1Task1['clock']['timestamp'] =
+        DateTime.now().add(Duration(hours: 1)).millisecondsSinceEpoch;
+
+    await devices[0].taskListService.mergeCrdtJson(jsonEncode(device1Crdt));
+    expect(
+      (await devices[0].taskListService.taskLists).toSet(),
+      {taskList1, taskList2},
+    );
+  });
+
+  test('crdt update less recent task in list', () async {
+    final taskList = TaskList(id: 'id', title: 'list');
+    final task1 = Task(id: 'task1id', title: 'task1');
+    final task2 = Task(id: task1.id, title: 'task2');
+    await devices[0].taskListService.upsertTaskList(taskList);
+    await devices[0].taskListService.upsertTask(taskList.id!, task1);
+    await Future.delayed(Duration(milliseconds: 10));
```

Comment:
Is this wait actually needed or could we instead manipulate a timestamp of a Task?

Reply:
Yeah, but this is easier and takes less lines of code. You actually shouldn't even need to wait because every operation takes time, this is just to be extra safe (in case it takes less than one ms). It also illustrates what this test is all about.

Reply:
If we don't need it, remove the wait completely.

If we do, let's test the logic and make it reproducible by explicitly adding a time delta to the second task.
More lines of code is neither bad nor good. If you think, the extra lines make the test more difficult to understand, create a helper method (e.g. `incrementTimestamp(task2, Duration(milliseconds: 1));`). This is explicit and we can reuse it in other tests as well.

Adding the delay introduces a code smell to our code base, which is likely to be copied by other developers when adding further tests.

Reply:
Removing it completely would, in theory, make it possible for the test to fail if it executes too quickly (less than 1ms between the insertion of both tasks).
The best I would agree to do is to change the delay to 1ms, which is the minimum for the test to be guaranteed to succeed.

**About manually adjusting the timestamp:**
It is currently not possible to  change the timestamp of tasks using the task list service because the service abstracts the underlying crdt implementation away. It is still be possible if you manipulate the json in the storage, but that is very hacky and totally dependent on the task list service implementation. This is done in the  'crdt merge clock drift' test where I even added comments that this is not a good test. I only did it there because it the only way to test this behavior on one device if you don't turn back your wall clock.
Another option would be to add a method to the task list service that allows for this timestamp manipulation. This would, however, only be useful in tests. Do you think this is worth the time investment?